tracer/src/Datadog.Trace/Agent/MessagePack/SpanModelBuilder.cs

```diff
@@ -0,0 +1,149 @@
+﻿// <copyright file="SpanModelBuilder.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace Datadog.Trace.Agent.MessagePack;
+
+internal struct SpanModelBuilder
+{
+    private readonly TraceChunkModel _traceChunk;
+
+    public readonly ulong LocalTraceRootSpanId;
+    public readonly bool LocalRootExists;
+
+    // for small trace chunks, use the ArraySegment<Span> copy directly, no heap allocations
+    private readonly ArraySegment<Span> _spans;
+
+    // for large trace chunks, use a HashSet<ulong> instead of iterating the array.
+    // note that his field is NOT readonly,
+    // but each copy of the same SpanModelBuilder can initialize it if needed.
+    private HashSet<ulong>? _hashSet;
+
+    public SpanModelBuilder(in TraceChunkModel traceChunk)
+    {
+        _traceChunk = traceChunk;
+        _spans = traceChunk.Spans;
+        _hashSet = null;
+
+        LocalTraceRootSpanId = traceChunk.LocalRootSpan?.SpanId ?? 0;
+        LocalRootExists = false;
+
+        // find the local root span to optimize for the common case of a "two-level" trace,
+        // i.e. a trace where all spans are direct descendants of the local root span, so there are no grandchildren,
+        // this makes it very fast to determine if a span's parent (the root) is present in the same trace chunk
+        if (LocalTraceRootSpanId > 0)
```

Comment:
🤔 In terms of responsibility, I think that knowing if a Local Root Exists in the trace chunk, knowing which span is the LocalTraceRootSpanId should be part of the `TraceChunkModel`. 

So I would actually move `LocalTraceRootSpanId`, `LocalRootExists`,  to the TraceChunkModel. If you agree, I don't mind this being done in a subsequent PR.

Reply:
I agree with the responsibility aspect, but right now the "analysis" of the trace chunk is done in `SpanModelBuilder`, which is called from `SpanMessagePackFormatter.Serialize()`, and the `TraceChunkModel` is already created at that point.

I was already thinking about `LocalRootExists` because there's another optimization we can make for the common case: if the trace chunk contains the distributed root span (`parentId=0`), we don't have to look for any other "chunk orphans." We know the agent will pick that one as the "chunk root."

I'll think about this some more!

Reply:
OK, I refactored this again :P

I moved the logic that analyses the spans into `TraceChunkModel` and we create this a little later now by pushing that down into `SpanBuffer.TryWrite()` for now.

`TraceChunkModel` now has `LocalRootSpanId` and `ContainsLocalRootSpan`.

Reply:
Hehe. I said you could do it in a different PR :p
thanks